File: backend/app/settlements/models/settlement.py
"""Settlement persistence models: money arriving, and money going out.

A settlement is one movement of money against a party -- a receipt from a
customer or a payment to a vendor -- together with the invoices it clears.

Both directions are one table because they are the same document with the
signs reversed. Splitting them would double every rule that matters (the
allocation cannot exceed the amount, the amount must reach the ledger, an
invoice cannot be over-cleared) and the two copies would drift, which is what
happened to the seven transactional document modules before they were given a
shared base.
"""

from datetime import date, datetime
from decimal import Decimal
from enum import StrEnum
from uuid import UUID

from sqlalchemy import (
    CheckConstraint,
    Date,
    ForeignKey,
    Index,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database.entity import BaseEntity
from app.core.database.types import UTCDateTime, UUIDType


class SettlementDirection(StrEnum):
    """Which way the money went."""

    RECEIPT = "RECEIPT"
    PAYMENT = "PAYMENT"
    #: Money back to a customer. It is money out like a payment and
    #: about a customer like a receipt, which is why it is neither of
    #: them: refunding a customer reduces what they have paid in
    #: advance rather than settling anything the firm owes a supplier.
    REFUND = "REFUND"


class SettlementMethod(StrEnum):
    """How the money moved, which decides the account it lands in."""

    CASH = "CASH"
    BANK = "BANK"


class SettlementStatus(StrEnum):
    """The lifecycle of a settlement.

    Two states, and no approval between them: a settlement is recorded after
    the money has moved, so there is nothing to decide. A mistake is taken back
    rather than edited -- the original stays, a mirror journal cancels it, and
    both are visible. Money that was recorded and then unrecorded is a fact
    about the day, not something to erase.
    """

    POSTED = "POSTED"
    REVERSED = "REVERSED"


class Settlement(BaseEntity):
    """Store one receipt from a customer or payment to a vendor."""

    __tablename__ = "settlements"
    __table_args__ = (
        UniqueConstraint(
            "firm_id", "settlement_number", name="UQ_settlements_firm_number"
        ),
        # Exactly one party, and it is the one the direction implies. A receipt
        # from a vendor is not a thing this document records.
        CheckConstraint(
            "(direction IN ('RECEIPT', 'REFUND') AND customer_id IS "
            "NOT NULL AND vendor_id IS NULL) OR (direction = 'PAYMENT' "
            "AND vendor_id IS NOT NULL AND customer_id IS NULL)",
            name="CK_settlements_party_matches_direction",
        ),
        CheckConstraint("amount > 0", name="CK_settlements_amount_positive"),
        Index("IX_settlements_firm_direction", "firm_id", "direction"),
        Index("IX_settlements_firm_date", "firm_id", "settlement_date"),
        Index("IX_settlements_firm_customer", "firm_id", "customer_id"),
        Index("IX_settlements_firm_vendor", "firm_id", "vendor_id"),
        Index("IX_settlements_firm_order", "firm_id", "sales_order_id"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    direction: Mapped[str] = mapped_column(String(20), nullable=False)
    customer_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("customers.id", ondelete="RESTRICT")
    )
    vendor_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("vendors.id", ondelete="RESTRICT")
    )
    settlement_number: Mapped[str] = mapped_column(String(60), nullable=False)
    settlement_date: Mapped[date] = mapped_column(Date, nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    #: What was allocated to invoices, and what was not. The remainder is an
    #: advance: money held against nothing in particular, which is a normal
    #: thing for a customer to send and has to be visible rather than inferred.
    allocated_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 2), nullable=False, default=Decimal("0.00"), server_default="0"
    )
    unallocated_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 2), nullable=False, default=Decimal("0.00"), server_default="0"
    )
    method: Mapped[str] = mapped_column(String(20), nullable=False)
    #: The cash or bank account the money actually moved through, resolved from
    #: the firm's control accounts at the time and then stored. Re-deriving it
    #: later would rewrite history if the mapping is ever changed.
    ledger_account_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("ledger_accounts.id", ondelete="RESTRICT"),
        nullable=False,
    )
    #: The order this money came in against, where it came in against one.
    #:
    #: A note about *why* the money arrived, not a ring-fence around it. The
    #: cash is the customer's balance either way: if the order is cancelled the
    #: deposit does not vanish, it stays on account. Recording the order is
    #: what makes "what has this customer paid us for order X" answerable, and
    #: what lets the bill for that order find the deposit when it is raised.
    sales_order_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("sales_orders.id", ondelete="RESTRICT")
    )
    instrument_reference: Mapped[str | None] = mapped_column(String(120))
    narration: Mapped[str | None] = mapped_column(Text())
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default=SettlementStatus.POSTED.value
    )
    #: The mirror journal that cancelled this one, and why. Set together with
    #: the status so a reversed settlement can always show what undid it.
    reversal_journal_entry_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("journal_entries.id", ondelete="RESTRICT")
    )
    reversed_at: Mapped[datetime | None] = mapped_column(UTCDateTime)
    reversed_by: Mapped[UUID | None] = mapped_column(UUIDType())
    reversal_reason: Mapped[str | None] = mapped_column(Text())
    #: The journal this wrote. A settlement that did not reach the ledger is
    #: the defect this module exists to fix, so the link is not optional.
    journal_entry_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("journal_entries.id", ondelete="RESTRICT"),
        nullable=False,
    )


class SettlementAllocation(BaseEntity):
    """Store how much of one settlement cleared one invoice."""

    __tablename__ = "settlement_allocations"
    __table_args__ = (
        CheckConstraint("amount > 0", name="CK_settlement_allocations_positive"),
        # One row per settlement per invoice: two lines against the same
        # invoice are one allocation, and keeping them apart would make the
        # outstanding arithmetic depend on how somebody typed it.
        UniqueConstraint(
            "settlement_id",
            "sales_invoice_id",
            name="UQ_settlement_allocations_sales_invoice",
        ),
        UniqueConstraint(
            "settlement_id",
            "purchase_invoice_id",
            name="UQ_settlement_allocations_purchase_invoice",
        ),
        Index("IX_settlement_allocations_sales", "firm_id", "sales_invoice_id"),
        Index("IX_settlement_allocations_purchase", "firm_id", "purchase_invoice_id"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    settlement_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("settlements.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    sales_invoice_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("sales_invoices.id", ondelete="RESTRICT")
    )
    purchase_invoice_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("purchase_invoices.id", ondelete="RESTRICT")
    )
    amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)

File: desktop/lib/models/settlement.dart
import 'entities.dart';

/// One invoice a settlement cleared, and by how much.
class SettlementAllocation {
  const SettlementAllocation({
    required this.id,
    required this.invoiceId,
    required this.invoiceNumber,
    required this.invoiceDate,
    required this.invoiceTotal,
    required this.amount,
  });

  final String id;
  final String invoiceId;
  final String invoiceNumber;
  final String invoiceDate;
  final String invoiceTotal;
  final String amount;

  factory SettlementAllocation.fromJson(Json json) => SettlementAllocation(
        id: stringValue(json['id']),
        invoiceId: stringValue(json['invoice_id']),
        invoiceNumber: stringValue(json['invoice_number']),
        invoiceDate: stringValue(json['invoice_date']),
        invoiceTotal: stringValue(json['invoice_total']),
        amount: stringValue(json['amount']),
      );
}

/// Money that arrived from a customer, or went out to a vendor.
class Settlement {
  const Settlement({
    required this.id,
    required this.direction,
    required this.partyId,
    required this.partyCode,
    required this.partyName,
    required this.settlementNumber,
    required this.settlementDate,
    required this.amount,
    required this.allocatedAmount,
    required this.unallocatedAmount,
    required this.method,
    required this.ledgerAccountName,
    required this.instrumentReference,
    required this.narration,
    required this.status,
    required this.journalEntryId,
    required this.reversalReason,
    required this.allocations,
    this.salesOrderNumber = '',
  });

  final String id;
  final String direction;
  final String partyId;
  final String partyCode;
  final String partyName;
  final String settlementNumber;
  final String settlementDate;
  final String amount;
  final String allocatedAmount;

  /// Money not tied to any invoice. It still reached the ledger and still
  /// reduced what the party owes in total; what it did not do is claim to have
  /// settled a particular document.
  final String unallocatedAmount;
  final String method;
  final String ledgerAccountName;
  final String instrumentReference;
  final String narration;
  final String status;

  /// The journal this wrote. Every settlement has one -- a settlement that did
  /// not reach the ledger is the thing the module exists to prevent.
  final String journalEntryId;
  final String reversalReason;
  final List<SettlementAllocation> allocations;

  /// The order this money came in against, where it came in against one. A
  /// note about why it arrived, not a ring-fence: cancelling the order does
  /// not make the deposit vanish.
  final String salesOrderNumber;

  /// Taken back. The original stays and a mirror journal cancels it, so a
  /// reversed settlement is still a record of money that arrived and was then
  /// unrecorded -- not an absence.
  bool get isReversed => status == 'REVERSED';

  bool get isOnAccount =>
      !isReversed && (double.tryParse(unallocatedAmount) ?? 0) > 0;

  factory Settlement.fromJson(Json json) {
    final Json d =
        json.containsKey('data') ? Map<String, dynamic>.from(json['data'] as Map) : json;
    final dynamic rows = d['allocations'];
    return Settlement(
      id: stringValue(d['id']),
      direction: stringValue(d['direction']),
      partyId: stringValue(d['party_id']),
      partyCode: stringValue(d['party_code']),
      partyName: stringValue(d['party_name']),
      settlementNumber: stringValue(d['settlement_number']),
      settlementDate: stringValue(d['settlement_date']),
      salesOrderNumber: stringValue(d['sales_order_number']),
      amount: stringValue(d['amount']),
      allocatedAmount: stringValue(d['allocated_amount']),
      unallocatedAmount: stringValue(d['unallocated_amount']),
      method: stringValue(d['method']),
      ledgerAccountName: stringValue(d['ledger_account_name']),
      instrumentReference: stringValue(d['instrument_reference']),
      narration: stringValue(d['narration']),
      status: stringValue(d['status']),
      journalEntryId: stringValue(d['journal_entry_id']),
      reversalReason: stringValue(d['reversal_reason']),
      allocations: [
        for (final dynamic row in rows is List ? rows : const [])
          if (row is Map)
            SettlementAllocation.fromJson(Map<String, dynamic>.from(row)),
      ],
    );
  }
}

/// One invoice with what is still owed on it.
class OutstandingInvoice {
  const OutstandingInvoice({
    required this.invoiceId,
    required this.invoiceNumber,
    required this.invoiceDate,
    required this.invoiceTotal,
    required this.allocatedAmount,
    required this.outstandingAmount,
  });

  final String invoiceId;
  final String invoiceNumber;
  final String invoiceDate;
  final String invoiceTotal;
  final String allocatedAmount;
  final String outstandingAmount;

  double get outstanding => double.tryParse(outstandingAmount) ?? 0;

  factory OutstandingInvoice.fromJson(Json json) => OutstandingInvoice(
        invoiceId: stringValue(json['invoice_id']),
        invoiceNumber: stringValue(json['invoice_number']),
        invoiceDate: stringValue(json['invoice_date']),
        invoiceTotal: stringValue(json['invoice_total']),
        allocatedAmount: stringValue(json['allocated_amount']),
        outstandingAmount: stringValue(json['outstanding_amount']),
      );
}

/// Spread an amount across invoices, oldest first.
///
/// This is what a cashier does by hand with a stack of invoices and a cheque,
/// and getting it wrong is tedious rather than interesting. Anything left over
/// when the invoices run out stays on account, which is a real outcome rather
/// than an error: a customer may well pay more than they currently owe.
Map<String, String> allocateOldestFirst(
  List<OutstandingInvoice> invoices,
  String amount,
) {
  double remaining = double.tryParse(amount) ?? 0;
  final Map<String, String> allocation = {};
  for (final OutstandingInvoice invoice in invoices) {
    if (remaining <= 0) break;
    final double take =
        remaining >= invoice.outstanding ? invoice.outstanding : remaining;
    if (take <= 0) continue;
    allocation[invoice.invoiceId] = take.toStringAsFixed(2);
    remaining -= take;
  }
  return allocation;
}
